src/Missings.jl

```diff
@@ -237,29 +237,30 @@ julia> collect(tx)
 """
 function skipmissings(args...)
     z = zip(args...)
-    return ntuple(i -> SkipMissings{typeof(z), i}(z, i), length(args))
+    return ntuple(i -> SkipMissings{typeof(z), i, typeof(z.is[i])}(z), length(args))
 end
 
-struct SkipMissings{Z, I}
+struct SkipMissings{Z, I, T}
     z::Z
-    k::Int
 end
-_inner_type(Z, I) = Z.parameters[1].parameters[I]
-_inner_itr(itr::SkipMissings) = itr.z.is[itr.k]
-_width(itr::SkipMissings) = length(first(itr.z))
+
+function _inner_itr(itr::SkipMissings{Z, I, T})::T  where {Z, I, T}
+    itr.z.is[I]
+end
+_width(itr::SkipMissings) = length(itr.z.is)
 _anymissing(itr::SkipMissings, i) = any(x -> x === missing, (y[i] for y in itr.z.is))
 const ZipofArrays = Base.Iterators.Zip{<:Tuple{Vararg{AbstractArray}}}
 
-function Base.show(io::IO, mime::MIME"text/plain", itr::SkipMissings{Z, I}) where {Z, I} 
-    show(io, mime, "SkipMissings{$Z, $I} based on $(_inner_itr(itr)) and $(_width(itr)-1) other iterators") 
+function Base.show(io::IO, mime::MIME"text/plain", itr::SkipMissings) 
+    show(io, mime, "SkipMissings based $(_inner_itr(itr)) and $(_length(itr.z.is)-1) other iterators") 
 end
 
 Base.IteratorSize(::Type{<:SkipMissings}) = Base.SizeUnknown()
-Base.IteratorEltype(::Type{SkipMissings{Z, I}}) where {Z, I} = Base.IteratorEltype(_inner_type(Z, I))
-Base.eltype(::Type{SkipMissings{Z, I}}) where {Z, I} = nonmissingtype(Base.eltype(_inner_type(Z, I)))
+Base.IteratorEltype(::Type{SkipMissings{Z, I, T}}) where {Z, I, T} = Base.IteratorEltype(T)
+Base.eltype(::Type{SkipMissings{Z, I, T}}) where {Z, I, T} = nonmissingtype(Base.eltype(T))
 
-function Base.iterate(itr::SkipMissings, state = 1)
-    y = iterate(itr.z, ntuple(_ -> state, _width(itr)))
+function Base.iterate(itr::SkipMissings{Z, I, T}, state = 1)::Tuple{eltype(T), Union{Int, Nothing}} where {Z, I, T}
```

Comment:
`@code_warntype sum(mx)` from your previous post seems to be OK if I remove this failing type assert:
```suggestion
function Base.iterate(itr::SkipMissings{Z, I, T}, state = 1) where {Z, I, T}
```
or if I fix it:
```suggestion
function Base.iterate(itr::SkipMissings{Z, I, T}, state = 1)::Union{Tuple{eltype(T), Int}, Nothing} where {Z, I, T}
```